python/tvm/relay/qnn/op/legalizations.py

```diff
@@ -0,0 +1,144 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, unused-argument
+"""Backend QNN related feature registration"""
+from __future__ import absolute_import
+
+import tvm
+from tvm import relay
+from .. import op as reg
+
+# Registering QNN Conv2D legalization function.
+@reg.register_qnn_legalize("qnn.conv2d")
+def legalize_qnn_conv2d(attrs, inputs, types):
+    """Legalizes QNN conv2d op.
+
+    Parameters
+    ----------
+    attrs : tvm.attrs.Attrs
+        Attributes of current convolution
+    inputs : list of tvm.relay.Expr
+        The args of the Relay expr to be legalized
+    types : list of types
+        List of input and output types
+
+    Returns
+    -------
+    result : tvm.relay.Expr
+        The legalized expr
+    """
+    return qnn_conv2d_legalize(attrs, inputs, types)
+
+# Generic QNN Conv2D legalization function.
+@tvm.target.generic_func
+def qnn_conv2d_legalize(attrs, inputs, types):
+    """Default legalization is None."""
+    return None
+
+# Intel x86 QNN Conv2D legalization function.
+@qnn_conv2d_legalize.register('cpu')
+def _qnn_conv2d_legalize(attrs, inputs, types):
+    """Legalizes QNN conv2d op. VNNI supports u8 x i8 fast conv/MM. If the dtypes are already good,
+    we dont transform. Else, we shift the tensor values and zero points to change the dtype.
+
+    Converting from int8 to uint8 can be done in following manner.
+
+    Original equation
+      scale * (QA - zp_a)
+      scale * (QA + 128 - 128 - zp_a)
+      scale * ( (QA + 128) - (zp_a + 128))
+
+    Replacing QA + 128 with QA' and (zp_a + 128) with zp_a'
+    We get our new quantized uint8 tensor - scale * (QA' - zp_a')
+
+    Similarly we can convert from int8 to uint8.
+
+    Parameters
+    ----------
+    attrs : tvm.attrs.Attrs
+        Attributes of current convolution
+    inputs : list of tvm.relay.Expr
+        The args of the Relay expr to be legalized
+    types : list of types
+        List of input and output types
+
+    Returns
+    -------
+    result : tvm.relay.Expr
+        The legalized expr
+    """
+
+    def _shift_to_uint8(data):
+        return _shift(data, 128, 'uint8')
+
+    def _shift_to_int8(data):
+        return _shift(data, -128, 'int8')
+
+    def _shift(data, shift, out_dtype):
+        """Shifts (add/subtracts) the qnn tensor with +/-128)"""
+        data_modified = relay.cast(data, 'int32')
+        data_modified = relay.add(data_modified, relay.const(shift, 'int32'))
+        data_modified = relay.cast(data_modified, out_dtype)
+        return data_modified
+
+    def _is_int8_hw_support(target):
+        """
```

Comment:
Can this be simplified to `return '-mcpu=skylake-avx512' in target.options`?

Or an easy-to-extend version

```python
def _support_vnni(target):
    supported_arches = {'-mcpu=skylake-avx512',}
    return supported_arches.intersection(set(target))
```